File: ibkr_show_frontend/src/api/charts.ts
import { request } from './http'
import type { EquityCurveResponse } from '@/types/charts'

export function fetchEquityCurve(params: {
  start_date?: string
  end_date?: string
} = {}): Promise<EquityCurveResponse> {
  const searchParams = new URLSearchParams()
  if (params.start_date) {
    searchParams.set('start_date', params.start_date)
  }
  if (params.end_date) {
    searchParams.set('end_date', params.end_date)
  }
  const queryString = searchParams.toString()
  return request<EquityCurveResponse>(`/api/charts/equity-curve${queryString ? `?${queryString}` : ''}`)
}

File: ibkr_show_backend/app/api/routes/charts.py
from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.api.deps import get_chart_service
from app.clients.es_client import ESClientError
from app.schemas.charts import EquityCurveResponse
from app.services.chart_service import ChartService

router = APIRouter(prefix="/charts", tags=["charts"])


@router.get("/equity-curve", response_model=EquityCurveResponse)
def get_equity_curve(
    start_date: str | None = Query(default=None),
    end_date: str | None = Query(default=None),
    service: ChartService = Depends(get_chart_service),
) -> EquityCurveResponse:
    try:
        return service.get_equity_curve(start_date=start_date, end_date=end_date)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)) from exc
    except ESClientError as exc:
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)) from exc
